evennia/contrib/pathfinding/tests.py

```diff
@@ -0,0 +1,211 @@
+from evennia import create_script, DefaultScript, DefaultRoom
+from evennia.contrib.pathfinding.pathfinder import Pathfinder
```

Comment:
Unittests are currently failing. This is because we don't include `networkx` as part of our dependencies. So while this may seem counter-intuitive, the `networkx` library needs to be mocked out for the unit tests. See [here](https://docs.python.org/3/library/unittest.mock-examples.html) for help with mocking.

Reply:
Counterintuitive indeed...I understand where you're coming from wrt not making nx a dependency and still wanting tests to pass, but having a hard time wrapping my mind around the rest of it. 

What becomes the point of the unit tests if we're faking the entire procedure? The current suite tests integration between libraries and will fail if Evennia or networkx deviate from each other in an incompatible manner. Mocking inputs/outputs seems like it would yield false assurances in such an event.

If mocking is just for the sake of passing tests in the absence of dependencies, I could much more easily catch an ImportError and skip all the tests, rather than write a lot of additional code that pretends to make sure everything's working in favor of code that actually does.

I don't mean to come across as defensive/combative, just feeling in over my head. I hear what you're asking but am struggling with comprehension of the concept :\

Reply:
@strikaco It may feel a bit unexpectedl, but it's worth remembering that we are not testing the networkx library. The pathfinder contrib has a lot of code that needs unit testing - this is the purpose of a unit test - to test the code in isolation. Mocking out a third-party library (for example because it requires a live network connection or, in this case because it's not necessarily available at the time of testing) is not weird at all. The networkx library has its own unit tests, it is not our job to unit test it - rather we expect a certain set of inputs and outputs from it. It's the *effect* of those inputs and outputs on the pathfinder code that we are testing, not the functionality of the networkx library itself.  

If you are worried about networkx's API changing in relation to Evennia you need to establish the specific version you support and lock into that. This is not unreasonable and something we do with actual default dependencies.

Of course the live networkx library is convenient to use when designing the unit test. But for consistency, there should be no use of the original networkx library in the final unit test suite (even if it's available) - the point being that the suite should be designed to act exactly the same regardless of if the inputs are coming from the networkx lib or from the mock. 

If you are unsure about how mocking works, I'm happy to help.

Reply:
> It's the effect of those inputs and outputs on the pathfinder code that we are testing, not the functionality of the networkx library itself.

@Griatch I'm still failing to understand how to apply this. What I'm envisioning is something like:

To test something like get_path(<basement obj>, <attic obj>)-- check to make sure after calling that, there is a subsequent call to nx.shortest_path(self, source=src_key, target=dst_key).

To test get_usable_path(<basement obj>, <attic obj>)-- check to make sure that after calling that, a request for a nx.DiGraph() is made, a bunch of calls are made to out_edges() and in_edges(), *.locks.check(), and *.add_edge().

This can't yield any output by itself. How much further am I supposed to go with each test? Is the entire point just to count/see what calls are being made?

Reply:
I've made an attempt to resolve this with 575737f; let me know if I'm on the right track...